packages/firestore/src/local/tab_notification_channel.ts

```diff
@@ -0,0 +1,393 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Code, FirestoreError } from '../util/error';
+import { BatchId, InstanceKey, TargetId } from '../core/types';
+import { assert } from '../util/assert';
+import { debug, error } from '../util/log';
+import { primitiveComparator } from '../util/misc';
+import { SortedSet } from '../util/sorted_set';
+import {
+  validateNamedArray,
+  validateNamedOptionalType,
+  validateNamedType
+} from '../util/input_validation';
+
+// Prefix keys used in WebStorage.
+const FIRESTORE_PREFIX = 'fs';
+const INSTANCE_ROW_PREFIX = 'instances';
+
+const LOG_TAG = 'TabNotificationChannel';
+
+/**
+ * WebStorage of the Firestore client.
+ *
+ * Firestore uses WebStorage for cross-tab notifications to propagate the
+ * state of pending mutation batches and query targets. All Firestore tabs are
+ * responsible for persisting their instance state as their in-memory state
+ * changes. Primary tabs will further update the state of each mutation batch
+ * and query target.
+ */
+export interface TabNotificationChannel {
+  /** Associates a new Mutation Batch with the current Firestore instance. */
+  addPendingMutation(batchId: BatchId): void;
```

Comment:
These names are getting a tad long, but I am not opposed. Done (used some slightly different names).

Reply:
Thanks. I hate to nit further, but I think you should use 'Global' and 'Local' instead of 'Globally' and 'Locally'.  'Global' and 'Local' are adjectives so they apply to the noun ('Mutation', 'QueryTarget', etc.) whereas 'Locally' and 'Globally' are adverbs so they apply to an adjective, and I don't know what "Locally active" vs "Globally active" means. Plus, I was hoping to use the same convention for all the methods.

That said, I'm feeling less gung-ho on the local vs global naming convention today, so if you want to drop it again, that's fine too (sorry).